File: app/src/main/java/es/etologic/mahjongscoring2/app/game/game_table/GameTableSeatsFragment.java
package es.etologic.mahjongscoring2.app.game.game_table;

import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.Locale;

import butterknife.BindColor;
import butterknife.BindDrawable;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;
import es.etologic.mahjongscoring2.R;
import es.etologic.mahjongscoring2.app.custom_views.VerticalTextView;
import es.etologic.mahjongscoring2.app.model.Seat;
import es.etologic.mahjongscoring2.app.model.SeatStates;
import es.etologic.mahjongscoring2.domain.model.enums.TableWinds;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;
import static es.etologic.mahjongscoring2.app.model.SeatStates.NORMAL;
import static es.etologic.mahjongscoring2.app.model.SeatStates.SELECTED;

public class GameTableSeatsFragment extends Fragment {

    //LISTENER
    interface TableSeatsListener {
        void onEastSeatClick();
        void onSouthSeatClick();
        void onWestSeatClick();
        void onNorthSeatClick();
    }

    //VIEWS
    @BindView(R.id.ivTableSeatEastSeatWindIcon) ImageView ivSeatEastWindIcon;
    @BindView(R.id.tvTableSeatEastName) TextView tvSeatEastName;
    @BindView(R.id.tvTableSeatEastPoints) TextView tvSeatEastPoints;
    @BindView(R.id.tvTableSeatEastPenaltyPoints) TextView tvSeatEastPenaltyPoints;
    @BindView(R.id.ivTableSeatSouthSeatWindIcon) ImageView ivSeatSouthWindIcon;
    @BindView(R.id.tvTableSeatSouthName) VerticalTextView tvSeatSouthName;
    @BindView(R.id.tvTableSeatSouthPoints) VerticalTextView tvSeatSouthPoints;
    @BindView(R.id.tvTableSeatSouthPenaltyPoints) VerticalTextView tvSeatSouthPenaltyPoints;
    @BindView(R.id.ivTableSeatWestSeatWindIcon) ImageView ivSeatWestWindIcon;
    @BindView(R.id.tvTableSeatWestName) TextView tvSeatWestName;
    @BindView(R.id.tvTableSeatWestPoints) TextView tvSeatWestPoints;
    @BindView(R.id.tvTableSeatWestPenaltyPoints) TextView tvSeatWestPenaltyPoints;
    @BindView(R.id.ivTableSeatNorthSeatWindIcon) ImageView ivSeatNorthWindIcon;
    @BindView(R.id.tvTableSeatNorthName) VerticalTextView tvSeatNorthName;
    @BindView(R.id.tvTableSeatNorthPoints) VerticalTextView tvSeatNorthPoints;
    @BindView(R.id.tvTableSeatNorthPenaltyPoints) VerticalTextView tvSeatNorthPenaltyPoints;
    //RESOURCES
    @BindDrawable(R.drawable.ic_east) Drawable eastIcon;
    @BindDrawable(R.drawable.ic_south) Drawable southIcon;
    @BindDrawable(R.drawable.ic_west) Drawable westIcon;
    @BindDrawable(R.drawable.ic_north) Drawable northIcon;
    @BindDrawable(R.drawable.ic_penalty_white) Drawable penaltyIcon;
    @BindDrawable(R.drawable.ic_trophy_white_18dp) Drawable tsumoIcon;
    @BindDrawable(R.drawable.ic_medal_white) Drawable ronIcon;
    @BindColor(android.R.color.white) int whiteColor;
    @BindColor(R.color.grayMM) int grayColor;
    @BindColor(R.color.colorAccent) int accentColor;
    @BindColor(R.color.red) int redColor;
    @BindColor(R.color.colorPrimary) int greenColor;
    //FIELDS
    private Unbinder unbinder;
    private TableSeatsListener listener;

    //EVENTS
    @OnClick(R.id.ivTableSeatEastSeatWindIcon) void onTableSeatEastSeatWindIconClick() { if (listener != null) listener.onEastSeatClick();}
    @OnClick(R.id.tvTableSeatEastName) void onTableSeatEastNameClick() { if (listener != null) listener.onEastSeatClick();}
    @OnClick(R.id.tvTableSeatEastPoints) void onTableSeatEastPointsClick() { if (listener != null) listener.onEastSeatClick();}
    @OnClick(R.id.tvTableSeatEastPenaltyPoints) void onTableSeatEastPenaltyPointsClick() { if (listener != null) listener.onEastSeatClick();}
    @OnClick(R.id.ivTableSeatSouthSeatWindIcon) void onTableSeatSouthSeatWindIconClick() { if (listener != null) listener.onSouthSeatClick();}
    @OnClick(R.id.tvTableSeatSouthName) void onTableSeatSouthNameClick() { if (listener != null) listener.onSouthSeatClick();}
    @OnClick(R.id.tvTableSeatSouthPoints) void onTableSeatSouthPointsClick() { if (listener != null) listener.onSouthSeatClick();}
    @OnClick(R.id.tvTableSeatSouthPenaltyPoints) void onTableSeatSouthPenaltyPointsClick() { if (listener != null) listener.onSouthSeatClick();}
    @OnClick(R.id.ivTableSeatWestSeatWindIcon) void onTableSeatWestSeatWindIconClick() { if (listener != null) listener.onWestSeatClick();}
    @OnClick(R.id.tvTableSeatWestName) void onTableSeatWestNameClick() { if (listener != null) listener.onWestSeatClick();}
    @OnClick(R.id.tvTableSeatWestPoints) void onTableSeatWestPointsClick() { if (listener != null) listener.onWestSeatClick();}
    @OnClick(R.id.tvTableSeatWestPenaltyPoints) void onTableSeatWestPenaltyPointsClick() { if (listener != null) listener.onWestSeatClick();}
    @OnClick(R.id.ivTableSeatNorthSeatWindIcon) void onTableSeatNorthSeatWindIconClick() { if (listener != null) listener.onNorthSeatClick();}
    @OnClick(R.id.tvTableSeatNorthName) void onTableSeatNorthNameClick() { if (listener != null) listener.onNorthSeatClick();}
    @OnClick(R.id.tvTableSeatNorthPoints) void onTableSeatNorthPointsClick() { if (listener != null) listener.onNorthSeatClick();}
    @OnClick(R.id.tvTableSeatNorthPenaltyPoints) void onTableSeatNorthPenaltyPointsClick() { if (listener != null) listener.onNorthSeatClick();}

    //PUBLIC METHODS
    void setTableSeatsListener(TableSeatsListener tableSeatsListener) {
        listener = tableSeatsListener;
    }

    void setEastSeat(Seat seat) {
        setWindIcon(ivSeatEastWindIcon, seat.getWind());
        setName(tvSeatEastName, seat.getName());
        setPoints(tvSeatEastPoints, seat.getPoints());
        setPenaltyPoints(tvSeatEastPenaltyPoints, seat.getPenalty());
        setState(ivSeatEastWindIcon, tvSeatEastName, tvSeatEastPoints, seat.getState());
    }
    private void setWindIcon(ImageView imageView, TableWinds wind) {
        switch (wind) {
            case EAST:
                imageView.setImageDrawable(eastIcon);
                break;
            case SOUTH:
                imageView.setImageDrawable(southIcon);
                break;
            case WEST:
                imageView.setImageDrawable(westIcon);
                break;
            case NORTH:
                imageView.setImageDrawable(northIcon);
                break;
            case NONE:
            default:
        }
    }
    private void setName(TextView textView, String name) {
        textView.setText(name);
    }
    private void setPoints(TextView textView, int points) {
        textView.setText(String.format(Locale.getDefault(), "%d", points));
    }
    private void setPenaltyPoints(TextView textView, int penaltyPoints) {
        textView.setText(String.format(Locale.getDefault(), "%+d", penaltyPoints));
        textView.setTextColor(penaltyPoints < 0 ? redColor : greenColor);
        textView.setVisibility(penaltyPoints != 0 ? VISIBLE : GONE);
    }
    private void setState(ImageView imageViewWind, TextView textViewName, TextView textViewPoints, SeatStates state) {
        if (state == NORMAL) {
            imageViewWind.clearColorFilter();
            textViewName.setTextColor(grayColor);
            textViewPoints.setTextColor(grayColor);
            imageViewWind.setEnabled(true);
            textViewName.setEnabled(true);
            textViewPoints.setEnabled(true);
        } else if (state == SELECTED) {
            imageViewWind.setColorFilter(accentColor);
            textViewName.setTextColor(accentColor);
            textViewPoints.setTextColor(accentColor);
            imageViewWind.setEnabled(true);
            textViewName.setEnabled(true);
            textViewPoints.setEnabled(true);
        } else {
            imageViewWind.setEnabled(false);
            textViewName.setEnabled(false);
            textViewPoints.setEnabled(false);
        }
    }
    void setSouthSeat(Seat seat) {
        setWindIcon(ivSeatSouthWindIcon, seat.getWind());
        setName(tvSeatSouthName, seat.getName());
        setPoints(tvSeatSouthPoints, seat.getPoints());
        setPenaltyPoints(tvSeatSouthPenaltyPoints, seat.getPenalty());
        setState(ivSeatSouthWindIcon, tvSeatSouthName, tvSeatSouthPoints, seat.getState());
    }
    void setWestSeat(Seat seat) {
        setWindIcon(ivSeatWestWindIcon, seat.getWind());
        setName(tvSeatWestName, seat.getName());
        setPoints(tvSeatWestPoints, seat.getPoints());
        setPenaltyPoints(tvSeatWestPenaltyPoints, seat.getPenalty());
        setState(ivSeatWestWindIcon, tvSeatWestName, tvSeatWestPoints, seat.getState());
    }
    void setNorthSeat(Seat seat) {
        setWindIcon(ivSeatNorthWindIcon, seat.getWind());
        setName(tvSeatNorthName, seat.getName());
        setPoints(tvSeatNorthPoints, seat.getPoints());
        setPenaltyPoints(tvSeatNorthPenaltyPoints, seat.getPenalty());
        setState(ivSeatNorthWindIcon, tvSeatNorthName, tvSeatNorthPoints, seat.getState());
    }
    //LIFECYCLE
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.game_table_seats_fragment, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        unbinder = ButterKnife.bind(this, view);
    }

    @Override
    public void onDestroyView() {
        if (unbinder != null) {
            unbinder.unbind();
        }
        super.onDestroyView();
    }
}

File: app/src/main/java/es/etologic/mahjongscoring2/app/game/game_table/GameTableSeatsFragment.kt
package es.etologic.mahjongscoring2.app.game.game_table

import android.R.color
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.core.content.ContextCompat
import es.etologic.mahjongscoring2.R
import es.etologic.mahjongscoring2.R.drawable
import es.etologic.mahjongscoring2.app.model.Seat
import es.etologic.mahjongscoring2.app.model.SeatStates
import es.etologic.mahjongscoring2.app.model.SeatStates.*
import es.etologic.mahjongscoring2.domain.model.enums.TableWinds
import kotlinx.android.synthetic.main.game_table_seat_east.*
import kotlinx.android.synthetic.main.game_table_seat_east.view.*
import kotlinx.android.synthetic.main.game_table_seat_north.*
import kotlinx.android.synthetic.main.game_table_seat_north.view.*
import kotlinx.android.synthetic.main.game_table_seat_south.*
import kotlinx.android.synthetic.main.game_table_seat_south.view.*
import kotlinx.android.synthetic.main.game_table_seat_west.*
import kotlinx.android.synthetic.main.game_table_seat_west.view.*
import java.util.*

class GameTableSeatsFragment : androidx.fragment.app.Fragment() {
    
    companion object {
        internal const val TAG = "GameTableSeatsFragment"
    }
    
    internal interface TableSeatsListener {
        fun onEastSeatClick()
        fun onSouthSeatClick()
        fun onWestSeatClick()
        fun onNorthSeatClick()
    }
    
    private var eastIcon: Drawable? = null
    private var southIcon: Drawable? = null
    private var westIcon: Drawable? = null
    private var northIcon: Drawable? = null
    private var penaltyIcon: Drawable? = null
    private var tsumoIcon: Drawable? = null
    private var ronIcon: Drawable? = null
    private var whiteColor: Int = 0
    private var grayColor: Int = 0
    private var accentColor: Int = 0
    private var redColor: Int = 0
    private var greenColor: Int = 0
    private var purplePenalty: Int = 0
    private var listener: TableSeatsListener? = null
    private var eastSeatState: SeatStates = NORMAL
    private var southSeatState: SeatStates = NORMAL
    private var westSeatState: SeatStates = NORMAL
    private var northSeatState: SeatStates = NORMAL
    
    //PUBLIC
    internal fun setTableSeatsListener(tableSeatsListener: TableSeatsListener) {
        listener = tableSeatsListener
    }
    
    internal fun setEastSeat(seat: Seat) {
        setWindIcon(ivTableSeatEastSeatWindIcon, seat.wind)
        setName(tvTableSeatEastName, seat.name)
        setPoints(tvTableSeatEastPoints, seat.points)
        setPenaltyPoints(tvTableSeatEastPenaltyPoints, seat.penalty)
        setState(ivTableSeatEastSeatWindIcon, tvTableSeatEastName, tvTableSeatEastPoints, tvTableSeatEastPenaltyPoints, seat.state)
        eastSeatState = seat.state
    }
    
    private fun setWindIcon(imageView: ImageView?, wind: TableWinds?) {
        when (wind) {
            TableWinds.EAST -> imageView?.setImageDrawable(eastIcon)
            TableWinds.SOUTH -> imageView?.setImageDrawable(southIcon)
            TableWinds.WEST -> imageView?.setImageDrawable(westIcon)
            TableWinds.NORTH -> imageView?.setImageDrawable(northIcon)
            else -> return
        }
    }
    
    private fun setName(textView: TextView, name: String?) {
        textView.text = name ?: ""
    }
    
    private fun setPoints(textView: TextView, points: Int) {
        textView.text = String.format(Locale.getDefault(), "%d", points)
    }
    
    private fun setPenaltyPoints(textView: TextView, penaltyPoints: Int) {
        textView.text = String.format(Locale.getDefault(), "%+d", penaltyPoints)
        textView.setTextColor(if (penaltyPoints < 0) redColor else greenColor)
        textView.visibility = if (penaltyPoints != 0) VISIBLE else GONE
    }
    
    private fun setState(
        imageViewWind: ImageView?,
        textViewName: TextView,
        textViewPoints: TextView,
        textViewPenaltyPoints: TextView,
        state: SeatStates?
    ) {
        if (state == NORMAL) {
            imageViewWind?.clearColorFilter()
            textViewName.setTextColor(grayColor)
            textViewPoints.setTextColor(grayColor)
            textViewPenaltyPoints.setTextColor(purplePenalty)
        } else if (state == SELECTED) {
            imageViewWind?.setColorFilter(accentColor)
            textViewName.setTextColor(accentColor)
            textViewPoints.setTextColor(accentColor)
            textViewPenaltyPoints.setTextColor(accentColor)
        }
    }
    
    internal fun setSouthSeat(seat: Seat) {
        setWindIcon(ivTableSeatSouthSeatWindIcon, seat.wind)
        setName(tvTableSeatSouthName, seat.name)
        setPoints(tvTableSeatSouthPoints, seat.points)
        setPenaltyPoints(tvTableSeatSouthPenaltyPoints, seat.penalty)
        setState(ivTableSeatSouthSeatWindIcon, tvTableSeatSouthName, tvTableSeatSouthPoints, tvTableSeatSouthPenaltyPoints, seat.state)
        southSeatState = seat.state
    }
    
    internal fun setWestSeat(seat: Seat) {
        setWindIcon(ivTableSeatWestSeatWindIcon, seat.wind)
        setName(tvTableSeatWestName, seat.name)
        setPoints(tvTableSeatWestPoints, seat.points)
        setPenaltyPoints(tvTableSeatWestPenaltyPoints, seat.penalty)
        setState(ivTableSeatWestSeatWindIcon, tvTableSeatWestName, tvTableSeatWestPoints, tvTableSeatWestPenaltyPoints, seat.state)
        westSeatState = seat.state
    }
    
    internal fun setNorthSeat(seat: Seat) {
        setWindIcon(ivTableSeatNorthSeatWindIcon, seat.wind)
        setName(tvTableSeatNorthName, seat.name)
        setPoints(tvTableSeatNorthPoints, seat.points)
        setPenaltyPoints(tvTableSeatNorthPenaltyPoints, seat.penalty)
        setState(ivTableSeatNorthSeatWindIcon, tvTableSeatNorthName, tvTableSeatNorthPoints, tvTableSeatNorthPenaltyPoints, seat.state)
        northSeatState = seat.state
    }
    
    //LIFECYCLE
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.game_table_seats_fragment, container, false)
        initResources()
        initListeners(view)
        return view
    }
    
    private fun initResources() {
        activity?.let {
            eastIcon = ContextCompat.getDrawable(it, drawable.ic_east)
            southIcon = ContextCompat.getDrawable(it, drawable.ic_south)
            westIcon = ContextCompat.getDrawable(it, drawable.ic_west)
            northIcon = ContextCompat.getDrawable(it, drawable.ic_north)
            penaltyIcon = ContextCompat.getDrawable(it, drawable.ic_penalty_white)
            tsumoIcon = ContextCompat.getDrawable(it, drawable.ic_trophy_white_18dp)
            ronIcon = ContextCompat.getDrawable(it, drawable.ic_medal_white)
            whiteColor = ContextCompat.getColor(it, color.white)
            grayColor = ContextCompat.getColor(it, R.color.grayMM)
            accentColor = ContextCompat.getColor(it, R.color.colorAccent)
            redColor = ContextCompat.getColor(it, R.color.red)
            greenColor = ContextCompat.getColor(it, R.color.colorPrimary)
            purplePenalty = ContextCompat.getColor(it, R.color.purplePenalty)
        }
    }
    
    private fun initListeners(view: View) {
        view.ivTableSeatEastSeatWindIcon?.setOnClickListener { if (eastSeatState != DISABLED) listener?.onEastSeatClick() }
        view.tvTableSeatEastName?.setOnClickListener { if (eastSeatState != DISABLED) listener?.onEastSeatClick() }
        view.tvTableSeatEastPoints?.setOnClickListener { if (eastSeatState != DISABLED) listener?.onEastSeatClick() }
        view.tvTableSeatEastPenaltyPoints?.setOnClickListener { if (eastSeatState != DISABLED) listener?.onEastSeatClick() }
        view.ivTableSeatSouthSeatWindIcon?.setOnClickListener { if (southSeatState != DISABLED) listener?.onSouthSeatClick() }
        view.tvTableSeatSouthName?.setOnClickListener { if (southSeatState != DISABLED) listener?.onSouthSeatClick() }
        view.tvTableSeatSouthPoints?.setOnClickListener { if (southSeatState != DISABLED) listener?.onSouthSeatClick() }
        view.tvTableSeatSouthPenaltyPoints?.setOnClickListener { if (southSeatState != DISABLED) listener?.onSouthSeatClick() }
        view.ivTableSeatWestSeatWindIcon?.setOnClickListener { if (westSeatState != DISABLED) listener?.onWestSeatClick() }
        view.tvTableSeatWestName?.setOnClickListener { if (westSeatState != DISABLED) listener?.onWestSeatClick() }
        view.tvTableSeatWestPoints?.setOnClickListener { if (westSeatState != DISABLED) listener?.onWestSeatClick() }
        view.tvTableSeatWestPenaltyPoints?.setOnClickListener { if (westSeatState != DISABLED) listener?.onWestSeatClick() }
        view.ivTableSeatNorthSeatWindIcon?.setOnClickListener { if (northSeatState != DISABLED) listener?.onNorthSeatClick() }
        view.tvTableSeatNorthName?.setOnClickListener { if (northSeatState != DISABLED) listener?.onNorthSeatClick() }
        view.tvTableSeatNorthPoints?.setOnClickListener { if (northSeatState != DISABLED) listener?.onNorthSeatClick() }
        view.tvTableSeatNorthPenaltyPoints?.setOnClickListener { if (northSeatState != DISABLED) listener?.onNorthSeatClick() }
    }
}
